app/controllers/courses_controller.rb

```diff
@@ -237,6 +239,32 @@ def wiki_params
       .permit(:language, :project)
   end
 
+  def courses_wikis_params
+    params
+      .require(:course)
+      .permit(wikis: [:language, :project])
+  end
+
+  def update_courses_wikis
+    @course.ensure_home_wiki_in_courses_wikis
+    multi_wikis = courses_wikis_params[:wikis]
+    return if multi_wikis.nil?
+
+    # Used to differentiate deleted wikis
+    old_wikis = @course.wikis.to_a
+    new_wikis = []
+
+    # Create an association for each new wiki
+    multi_wikis.each do |wiki|
+      new_wiki = Wiki.get_or_create(language: wiki[:language], project: wiki[:project])
+      new_wikis.push(new_wiki)
+      @course.wikis.push(new_wiki) unless old_wikis.include? new_wiki
+    end
+
+    # Delete removed wikis except home wiki
+    @course.wikis.delete(old_wikis - new_wikis - [@course.home_wiki])
```

Comment:
There's probably a simpler way to do the adding and removing. ActiveRecord has a lot of support for join models like `CoursesWikis`, such that you can do something like `@course.wikis = set_of_wikis`, and it create or delete records as needed to make that happen.

Reply:
You can also do `@course.update(wikis: set_of_wikis)` and will do the same thing.

So maybe take the wiki params, map them to a set of actual Wiki records with `Wiki.get_or_create`, and then do the update.

Reply:
Hmm... thinking more on this along with the `ensure_home_wiki_in_courses_wikis`, here's what I think should happen...

Since changing the home_wiki or any other wikis will only happen through `CoursesController#update`, and you already have some special handling of it, the actual logic for updating the wikis could be moved to `Course` itself, with a new method like `#update_wikis`... which could do a `update(wikis: updated_wikis)` followed by `ensure_home_wiki_in_courses_wikis`... and then `ensure_home_wiki_in_courses_wikis` could just be a `before_create` callback instead of `before_save`. That way it would not add any extra queries to the updates that happen in other contexts.